### PR TITLE
Data binding for HTML components 3288

### PR DESCRIPTION
**Pull request by ilushaaa**
#3288

### CODE REVIEW

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        void remove();
+    }
+
+    class HtmlContainerTextFormatter implements Formatter<Object> {
+
+        @Override
+        public String apply(@Nullable Object value) {
+            if (value instanceof Collection<?>) {
+                return formatCollection((Collection<?>) value);
+            }
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java:154]: image


Java 17 baby:
```suggestion
            if (value instanceof Collection<?> collectionValue) {
                return formatCollection(collectionValue);
            }
```

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/html/AbstractHtmlContainerLoader.java [hunk 2]
```diff
@@ -40,5 +49,25 @@ public void loadComponent() {
         componentLoader().loadWhiteSpace(resultComponent, element);
         componentLoader().loadSizeAttributes(resultComponent, element);
         componentLoader().loadThemeList(resultComponent, element);
+
+        String property = element.attributeValue("property");
+        if (!Strings.isNullOrEmpty(property) && resultComponent != null) {
+            Optional<InstanceContainer<?>> instanceContainer = getDataLoaderSupport().loadContainer(element, property);
+            instanceContainer.ifPresent(container -> getHtmlDataBinding().bind(resultComponent, container, property));
+        }
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/html/AbstractHtmlContainerLoader.java:57]: 1. I'm sure we should move this into a separate `loadData(resultComponent, element)` method.
2. Unnecessary variable `instanceContainer` it seems:

```suggestion
getDataLoaderSupport().loadContainer(element, property)
            .ifPresent(container -> getHtmlDataBinding().bind(resultComponent, container, property));
```

3. Result component can be null? WAT? 54 line

File: jmix-flowui/flowui/src/test/groovy/data_binding_html_container/HtmlContainerReadonlyDataBindingTest.groovy [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.jdbc.core.JdbcTemplate
+import test_support.entity.Animal
+import test_support.entity.Zoo
+import test_support.spec.FlowuiTestSpecification
+
+import java.util.stream.Collectors
+
+@SpringBootTest
+class HtmlContainerReadonlyDataBindingTest extends FlowuiTestSpecification {
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/test/groovy/data_binding_html_container/HtmlContainerReadonlyDataBindingTest.groovy:35]: I think we should write one more test for a HTML component that only has a `property` attribute, and this HTML component should be located in a `formLayout` in which the `dataContainer` is specified

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/html/AbstractHtmlContainerLoader.java [hunk 4]
```diff
@@ -40,5 +49,25 @@ public void loadComponent() {
...
+    protected DataLoaderSupport getDataLoaderSupport() {
+        if (dataLoaderSupport == null) {
+            dataLoaderSupport = applicationContext.getBean(DataLoaderSupport.class, context);
+        }
+        return dataLoaderSupport;
+    }
+
+    protected HtmlContainerReadonlyDataBinding getHtmlDataBinding() {
+        if (htmlComponentReadonlyDataBinding == null) {
+            htmlComponentReadonlyDataBinding = applicationContext.getBean(HtmlContainerReadonlyDataBinding.class, context);
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/html/AbstractHtmlContainerLoader.java:69]: You don't need to pass the `context` as an additional variable
`DataLoaderSupport` bean requires `context` as a constructor argument

In the `HtmlContainerReadonlyDataBinding` constructor, you are requests for a bean, which will autowired automatically
There's definitely no need for context there

```suggestion
            htmlComponentReadonlyDataBinding = applicationContext.getBean(HtmlContainerReadonlyDataBinding.class);
```

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+@Component("flowui_HtmlContainerDataBinding")
+public class HtmlContainerReadonlyDataBindingImpl implements HtmlContainerReadonlyDataBinding {
+
+    protected MetadataTools metadataTools;
+
+    public HtmlContainerReadonlyDataBindingImpl(MetadataTools metadataTools) {
+        this.metadataTools = metadataTools;
+    }
+
+    protected List<Pair<HtmlContainer, HtmlContainerListenerRegistrationsCleaner>> listenersRegistrations = new LinkedList<>();
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java:47]: More than 120 characters per line

List of a pairs... Maybe we need to use Map?

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    @Override
+    public void bind(HtmlContainer htmlContainer, ValueSource<?> valueSource, Formatter<Object> formatter) {
+        Preconditions.checkNotNullArgument(htmlContainer);
+        Preconditions.checkNotNullArgument(valueSource);
+        Preconditions.checkNotNullArgument(formatter);
+
+        updateHtmlContainerText(htmlContainer, formatter, valueSource.getValue());
+
+        Registration registration = valueSource.addValueChangeListener(valueChangeEvent ->
+            updateHtmlContainerText(htmlContainer, formatter, valueChangeEvent.getValue()));
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java:62]: Not aligned (ALT-CTRL-L)

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    protected void updateHtmlContainerText(HtmlContainer htmlContainer,
+                                           Formatter<Object> formatter,
+                                           @Nullable Object value) {
+        htmlContainer.setText(formatter.apply(value));
+    }
+
+    @FunctionalInterface
+    protected interface HtmlContainerListenerRegistrationsCleaner {
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java:144]: We don't need this, we can use `com.vaadin.flow.shared.Registration`

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+    protected MetadataTools metadataTools;
+
+    public HtmlContainerReadonlyDataBindingImpl(MetadataTools metadataTools) {
+        this.metadataTools = metadataTools;
+    }
+
+    protected List<Pair<HtmlContainer, HtmlContainerListenerRegistrationsCleaner>> listenersRegistrations = new LinkedList<>();
+
+    public void bind(HtmlContainer htmlComponent, ValueSource<?> valueSource) {
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java:49]: What happens if I try to bind a component to one container and then to another one?

I don't see a way to automatically unbind a container if it has already been bound to any one.
I think that's why we should use a map to persist subscriptions.

Then, when we try to bind a new container, we will have to unbind the old one, write in the log that we unbind
And bind a new one
See `io.jmix.flowui.component.delegate.AbstractValueComponentDelegate#setValueSource` as a source of inspiration

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        bind(htmlComponent, valueSource, new HtmlContainerTextFormatter());
+    }
+
+    @Override
+    public void bind(HtmlContainer htmlContainer, ValueSource<?> valueSource, Formatter<Object> formatter) {
+        Preconditions.checkNotNullArgument(htmlContainer);
+        Preconditions.checkNotNullArgument(valueSource);
+        Preconditions.checkNotNullArgument(formatter);
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java:57]: We use this quite often, maybe we need static import?

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                                           @Nullable Object value) {
+        htmlContainer.setText(formatter.apply(value));
+    }
+
+    @FunctionalInterface
+    protected interface HtmlContainerListenerRegistrationsCleaner {
+        void remove();
+    }
+
+    class HtmlContainerTextFormatter implements Formatter<Object> {
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java:148]: I don't like the concept of a non-static class that uses one `MetadataTools` bean from its parent
I think we should remove it

Moreover, the parent class doesn't use the bean at all O_O

You have a clear division between collection binding and instance binding
For collection binding, you can use the ready-made `io.jmix.flowui.component.formatter.CollectionFormatter`
And for a instance one, you can simply use a lambda with `MetadataTools#apply`

**Review by KremnevDmitry — CHANGES_REQUESTED** [review]

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+import static io.jmix.core.common.util.Preconditions.checkNotEmptyString;
+import static io.jmix.core.common.util.Preconditions.checkNotNullArgument;
+
+@Component("flowui_HtmlContainerDataBinding")
+public class HtmlContainerReadonlyDataBindingImpl implements HtmlContainerReadonlyDataBinding {
+
+    private static final Logger log = LoggerFactory.getLogger(HtmlContainerReadonlyDataBindingImpl.class);
+
+    protected MetadataTools metadataTools;
+    protected List<Pair<HtmlContainer, DataBindingListenerCleaner>> listenersRegistrations = new LinkedList<>();
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java:50]: Use `Map<HtmlContainer, Runnable>`

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+        listenersRegistrations.add(new Pair<>(htmlContainer, propertyChangeCleaner));
+
+        Subscription itemChangeSubscription = dataContainer.addItemChangeListener(itemChangeEvent -> {
+            Object propertyValue = itemChangeEvent.getItem() != null
+                    ? EntitySystemAccess.getEntityEntry(itemChangeEvent.getItem()).getAttributeValue(property)
+                    : null;
+            updateHtmlContainerText(htmlContainer, formatter, propertyValue);
+        });
+        DataBindingListenerCleaner itemChangeCleaner = itemChangeSubscription::remove;
+        listenersRegistrations.add(new Pair<>(htmlContainer, itemChangeCleaner));
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java:115]: When `list` will be replaced to `map` the following code will be applicable

```java
        listenersRegistrations.put(htmlContainer, () -> {
            propertyChangeSubscription.remove();
            itemChangeSubscription.remove();
        });
```

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+            if (htmlContainer.equals(htmlContainerAndCleaner.getFirst())) {
+                htmlContainerAndCleaner.getSecond().remove();
+                iterator.remove();
+
+                if (warnIfBound) {
+                    log.warn("Silent unbind. HtmlContainer with id = '{}' was already bounden with data source.",
+                            htmlContainer.getId());
+                }
+            }
+        }
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java:172]: When `list` will be replaced to `map` then you can simply use the following construction here:

```suggestion
        Runnable remove = listenersRegistrations.get(htmlContainer);
        if (remove == null) {
            return;
        }

        remove.run();
        listenersRegistrations.remove(htmlContainer);
        
        if (warnIfBound) {
            log.warn("Silent unbind. HtmlContainer with id = '{}' was already bounden with data source.",
                    htmlContainer.getId());
        }
```

**Review by KremnevDmitry — CHANGES_REQUESTED** [review]

**Review by KremnevDmitry — APPROVED** [review]
Allowed to review by GG after aligning the lines and checking security permissions

**Review by glebfox — CHANGES_REQUESTED** [review]

**Review by glebfox — APPROVED** [review]

**Pull request merged**